src/main/java/edu/hm/hafner/analysis/ast/CheckStyleAstFactory.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.Arrays;
+
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Factory to pick the best AST implementation based on the type of the CheckStyle warning.
+ *
+ * @author Christian Möstl
+ */
+// CHECKSTYLE:COUPLING-OFF
+public final class CheckStyleAstFactory {
+    private static final String[] METHOD_AST = {"CovariantEquals", "EqualsAvoidNull",
+            "EqualsHashCode", "HiddenField", "JavadocMethod", "MethodLength", "MethodName", "MethodParamPad",
+            "MethodTypeParameterName", "ParameterAssignment", "ParameterName", "ParameterNumber", "SuperClone",
+            "SuperFinalize", "ThrowsCount"};
+    private static final String[] SURROUNDING_ELEMENTS_AST = {"AnonInnerLength", "ArrayTypeStyle", "AvoidNestedBlocks",
+            "BooleanExpressionComplexity", "DefaultComesLast", "EmptyBlock", "EmptyForIteratorPad", "EmptyStatement",
+            "FallThrough", "FileTabCharacter", "GenericWhitespace", "IllegalCatch", "IllegalThrows", "InnerAssignment", "JavaNCSS",
+            "LeftCurly", "LocalFinalVariableName", "LocalVariableName", "MissingSwitchDefault",
+            "ModifiedControlVariable", "MultipleVariableDeclarations", "NeedBraces", "NestedForDepth", "NestedIfDepth",
+            "NestedTryDepth", "NoWhitespaceAfter", "NoWhitespaceBefore", "NPathComplexity", "OneStatementPerLine",
+            "OperatorWrap", "ParenPad", "RightCurly", "SimplifyBooleanExpression", "SimplifyBooleanReturn",
+            "StringLiteralEquality", "TypecastParenPad", "UnnecessaryParentheses", "UpperEll", "WhitespaceAfter",
+            "WhitespaceAround"};
+    private static final String[] FILE_AST = {"ClassDataAbstractionCoupling", "ClassFanOutComplexity",
+            "FileLength", "IllegalImport", "InterfaceIsType", "OuterTypeFilename", "PackageDeclaration"};
+    private static final String[] CLASS_AST = {"ClassTypeParameterName", "FinalClass",
+            "HideUtilityClassConstructor", "InnerTypeLast", "JavadocType", "JUnitTestCase", "MultipleStringLiterals",
+            "MutableException", "TypeName"};
+    private static final String[] METHOD_OR_CLASS_AST = {"AnnotationUseStyle",
+            "JavadocStyle", "MissingDeprecated", "ModifierOrder", "RedundantModifier", "VisibilityModifier"};
+    private static final String[] FIELDS_AST = {"ConstantName", "ExplicitInitialization",
+            "JavadocVariable", "MemberName", "StaticVariableName"};
+    private static final String[] NAME_PACKAGE_AST = {"PackageName"};
+
+    /**
+     * Returns an instance of an {@link Ast} specific for the specified warning type.
+     *
+     * @param filename    the file name of the Java class that contains the warning
+     * @param warningType the type of the warning
+     * @param lineNumber  the primary line number of the warning
+     * @return the specific ast
+     */
+    @SuppressWarnings("PMD.CyclomaticComplexity")
+    public static Ast getInstance(final String filename, final String warningType, final int lineNumber) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [getInstance method always creates a new object and hence does not comply to Singleton Design Pattern behaviour. Please review](https://app.codacy.com/manual/uhafner/analysis-model/pullRequest?prid=5019963)